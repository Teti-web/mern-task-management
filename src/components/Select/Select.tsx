import { FC, useState } from "react";
import { ChevronDownIcon } from "@heroicons/react/24/outline";

export type SelectType = {
  selectName: string;
  isBorder?: boolean;
  options: {
    value: string;
    text: string;
  }[];
  onChange: (value: string) => void;
};

const Select: FC<SelectType> = ({
  options,
  onChange,
  selectName,
  isBorder,
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const [selectedValue, setSelectedValue] = useState<string>(selectName);

  let buttonSelectClass =
    "flex justify-between items-center gap-2 text-secondary-500 w-full text-left font-primary font-medium text-[12px] px-[10px] py-[14px] transition-all ease-in-out duration-300 md:px-[16px] md:py-[20px] lg:text-[14px] hover:text-primary-500 focus:text-primary-600";
  const chevronClass = `w-[20px] h-[20px] text-secondary-500 transition-transform duration-300 ${isOpen ? "transform -rotate-180" : ""}`;

  if (isBorder) {
    buttonSelectClass +=
      " border-[1px] border-solid rounded-[10px] border-border bg-white hover:border-primary-100 hover:text-secondary-500 focus:border-primary-200 focus:text-secondary-500";
  }

  return (
    <div className="w-full relative">
      <button className={buttonSelectClass} onClick={() => setIsOpen(!isOpen)}>
        {selectedValue}
        <ChevronDownIcon className={chevronClass} />
      </button>
      {isOpen && (
        <ul className=" absolute grid gap-3 w-full z- border-solid border-[1px] rounded-[10px] border-border bg-white px-[10px] py-[14px] leading-snug font-primary font-medium text-[12px] md:px-[16px] md:py-[20px] lg:text-[14px]">
          {options.map((option, index) => (
            <li
              className="cursor-pointer hover:text-primary-500"
              key={index}
              onClick={() => {
                setSelectedValue(option.text);
                onChange(option.value);
                setIsOpen(false);
              }}
            >
              {option.text}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default Select;
