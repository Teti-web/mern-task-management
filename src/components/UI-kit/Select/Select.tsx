import { FC, useState } from "react";
import { ChevronDownIcon } from "@heroicons/react/24/outline";

export type SelectType = {
  buttonValue: string;
  isBorder?: boolean;
  options: {
    value: string;
    text: string;
  }[];
  onChange: (value: string) => void;
};

const Select: FC<SelectType> = ({
  options,
  onChange,
  isBorder,
  buttonValue,
}) => {
  const [isOpen, setIsOpen] = useState(false);

  let buttonSelectClass =
    "flex justify-between items-center gap-[3px] text-secondary-500 w-full text-left font-primary font-medium text-[12px] px-[4px] py-[6px] transition-all ease-in-out duration-300 lg:text-[14px] hover:text-primary-500 focus:text-primary-600";
  const chevronClass = `w-[20px] h-[20px] text-secondary-500 transition-transform duration-300 ${isOpen ? "transform -rotate-180" : ""}`;

  if (isBorder) {
    buttonSelectClass +=
      " px-[10px] py-[14px] md:px-[16px] md:py-[20px] border-[1px] border-solid rounded-[10px] border-border bg-white hover:border-primary-100 hover:text-secondary-500 focus:border-primary-200 focus:text-secondary-500";
  }

  const handleLanguageChange = (option: { value: string; text: string }) => {
    onChange(option.value);
    setIsOpen(false);
  };

  return (
    <div className="w-full relative">
      <button className={buttonSelectClass} onClick={() => setIsOpen(!isOpen)}>
        {buttonValue || options[0].text}{" "}
        <ChevronDownIcon className={chevronClass} />
      </button>
      {isOpen && (
        <ul className=" absolute grid gap-3 w-full z- border-solid border-[1px] rounded-[10px] border-border bg-white px-[10px] py-[14px] leading-snug font-primary font-medium text-[12px] md:px-[16px] md:py-[20px] lg:text-[14px]">
          {options.map((option, index) => (
            <li
              className="cursor-pointer hover:text-primary-500"
              key={index}
              onClick={() => handleLanguageChange(option)}
            >
              {option.text}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default Select;
