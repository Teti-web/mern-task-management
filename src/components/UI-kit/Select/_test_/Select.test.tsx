import "@testing-library/jest-dom";
import { render, cleanup, fireEvent } from "@testing-library/react";

import Select from "../Select";

describe("Select componet", () => {
  afterEach(cleanup);

  const options = [
    { value: "en", text: "English" },
    { value: "fr", text: "French" },
    { value: "es", text: "Spanish" },
  ];

  it("renders Select component with default value", () => {
    const onChange = jest.fn();
    const { getByText, asFragment } = render(
      <Select options={options} onChange={onChange} buttonValue={""} />
    );
    const defaultOption = getByText("English");
    expect(defaultOption).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders Select component with provided button value", () => {
    const onChange = jest.fn();
    const { getByText, asFragment } = render(
      <Select options={options} onChange={onChange} buttonValue="French" />
    );
    const selectedOption = getByText("French");
    expect(selectedOption).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls onChange when an option is clicked", () => {
    const onChange = jest.fn();
    const { getByText, asFragment } = render(
      <Select options={options} onChange={onChange} buttonValue={""} />
    );

    fireEvent.click(getByText("English"));

    const optionToClick = getByText("French");
    fireEvent.click(optionToClick);
    expect(onChange).toHaveBeenCalledWith("fr");
    expect(asFragment()).toMatchSnapshot();
  });
  it("opens and closes the dropdown when button is clicked", () => {
    const onChange = jest.fn();
    const { getByText, queryByText, asFragment } = render(
      <Select options={options} onChange={onChange} buttonValue={""} />
    );
    const button = getByText("English");
    fireEvent.click(button);
    const dropdownOption = getByText("Spanish");
    expect(dropdownOption).toBeInTheDocument();
    fireEvent.click(button);
    const closedOption = queryByText("Spanish");
    expect(closedOption).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
