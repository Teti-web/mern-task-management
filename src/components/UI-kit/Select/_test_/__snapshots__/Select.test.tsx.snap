// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select componet calls onChange when an option is clicked 1`] = `
<DocumentFragment>
  <div
    class="w-full relative"
  >
    <button
      class="flex justify-between items-center gap-[3px] text-secondary-500 w-full text-left font-primary font-medium text-[12px] px-[4px] py-[6px] transition-all ease-in-out duration-300 lg:text-[14px] hover:text-primary-500 focus:text-primary-600"
    >
      English 
      <svg
        aria-hidden="true"
        class="w-[20px] h-[20px] text-secondary-500 transition-transform duration-300 "
        data-slot="icon"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m19.5 8.25-7.5 7.5-7.5-7.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Select componet opens and closes the dropdown when button is clicked 1`] = `
<DocumentFragment>
  <div
    class="w-full relative"
  >
    <button
      class="flex justify-between items-center gap-[3px] text-secondary-500 w-full text-left font-primary font-medium text-[12px] px-[4px] py-[6px] transition-all ease-in-out duration-300 lg:text-[14px] hover:text-primary-500 focus:text-primary-600"
    >
      English 
      <svg
        aria-hidden="true"
        class="w-[20px] h-[20px] text-secondary-500 transition-transform duration-300 "
        data-slot="icon"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m19.5 8.25-7.5 7.5-7.5-7.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Select componet renders Select component with default value 1`] = `
<DocumentFragment>
  <div
    class="w-full relative"
  >
    <button
      class="flex justify-between items-center gap-[3px] text-secondary-500 w-full text-left font-primary font-medium text-[12px] px-[4px] py-[6px] transition-all ease-in-out duration-300 lg:text-[14px] hover:text-primary-500 focus:text-primary-600"
    >
      English 
      <svg
        aria-hidden="true"
        class="w-[20px] h-[20px] text-secondary-500 transition-transform duration-300 "
        data-slot="icon"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m19.5 8.25-7.5 7.5-7.5-7.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Select componet renders Select component with provided button value 1`] = `
<DocumentFragment>
  <div
    class="w-full relative"
  >
    <button
      class="flex justify-between items-center gap-[3px] text-secondary-500 w-full text-left font-primary font-medium text-[12px] px-[4px] py-[6px] transition-all ease-in-out duration-300 lg:text-[14px] hover:text-primary-500 focus:text-primary-600"
    >
      French 
      <svg
        aria-hidden="true"
        class="w-[20px] h-[20px] text-secondary-500 transition-transform duration-300 "
        data-slot="icon"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m19.5 8.25-7.5 7.5-7.5-7.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
